lib/options_validator.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+const isObject = require('./utils').isObject;
+
+let validationLevel = 'warn';
+
+function createOptionsSchema(optionsSchema) {
+  return validationFunction.bind(optionsSchema);
+}
+
+function validationFunction(providedOptions) {
+  let validatedOptions = Object.assign({}, providedOptions);
+
+  Object.keys(validatedOptions).forEach(option => {
+    let optionValue = validatedOptions[option];
+    let optionType = this[option].type;
+
+    let invalidateMessage = `${option} should be of type ${optionType}, but is of type ${typeof optionValue}.`;
+
+    if (typeof optionType === 'string') {
+      if (optionType === 'object') {
+        if (!isObject(optionValue)) {
+          invalidate(validationLevel, invalidateMessage);
+        }
+      }
+      if (typeof optionValue !== optionType) {
+        invalidate(validationLevel, invalidateMessage);
+      }
+    } else if (Array.isArray(optionType)) {
+      if (optionType.indexOf(typeof optionValue) === -1) {
+        invalidate(validationLevel, invalidateMessage);
+      }
+    } else {
+      if (!(optionValue instanceof optionType)) {
+        invalidate(validationLevel, invalidateMessage);
+      }
+    }
+  });
+  return Object.freeze(validatedOptions);
+}
+
+function invalidate(validationLevel, message) {
+  if (validationLevel === 'warn') {
+    console.warn(message);
+  } else if (validationLevel === 'error') {
+    throw new Error(message);
+  }
+}
+
+function setValidationLevel(providedValidationLevel) {
```

Comment:
do we really need a `setValidationLevel`? Presumably this is to change the behavior during runtime? First off, I think we should avoid providing a setter for a globally scoped variable in the module. Secondly, I think we should just build the validator function with a fixed validation level - do I need to toggle this during runtime?

Reply:
Like we talked about yesterday, a better way to set the `validationLevel` would be to make `validationLevel` a parameter of `validationFunction`.

I could set a `validationLevel` on a per-class basis, so `Collection` would have a `validationLevel` associated with it and each operation in `Collection` would use that `validationLevel` as an input to its `validationFunction`. Then I would need to provide a `setValidationLevel` function on each class because the user needs to be able to specify the `validationLevel`.

Alternatively, I could make `validationLevel` an option for `MongoClient`. `MongoClient`'s `validationLevel` could be set by the user and all other classes, like `Collection`, would inherit that `validationLevel`.

I think the second option makes the most sense because users would likely want the same `validationLevel` for all operations and classes.

@daprahamian and @mbroadst, what do you think?

Reply:
I imagine in the most common case we will be using the `OperationBuilder`, and you can work support for passing in a validation level from the current context (inherited property for example) there, however, for the basic production of a validation function I think the only way forward is to pass this in as an option. I think your validation function creator should result in the following signature: `fn(objectToValidate, options)`, and one of those options should be `validationLevel`